### PR TITLE
Fix data source info is missing in default query when click Discover from other pages

### PR DESCRIPTION
**Pull request by gaobinlong**
### Description

When we click Discover sidebar from other pages, we don't fetch the data source info and enrich the default dataset, so that some unexpected behavior happens.

### Issues Resolved

https://github.com/opensearch-project/OpenSearch-Dashboards/issues/8582

## Screenshot

No UI change.

The direct impact is that when click `Discover` from other pages, the selected default dataset always shows the data source prefix, and makes other components work.

image

## Testing the changes
1. Create an index pattern in an external data source, and set it as the default dataset.
2. Click `Discover` from other pages, check if the selected dataset shows the data source prefix, and also check if the url contains data source info.

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [x] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by gaobinlong** [comment]
@kavilla @ashwin-pc could you help to review this PR? Thank you!

**Comment by gaobinlong** [comment]
@SuZhou-Joe @ruanyl could you help to review this PR, thank you!

File: src/plugins/data/public/query/query_string/dataset_service/dataset_service.ts [hunk 1]
```diff
@@ -227,15 +232,31 @@ export class DatasetService {
       return undefined;
     }
 
+    let dataSource;
+    if (indexPattern.dataSourceRef) {
+      dataSource = await this.savedObjectsClient?.get(
```
Comment (ruanyl) [inline src/plugins/data/public/query/query_string/dataset_service/dataset_service.ts:237]: @gaobinlong You can use `this.indexPatterns?.getDataSource(indexPattern.dataSourceRef?.id)` to simplify the implementation. Then passing `savedObjectsClient` is unnecessary.

Comment (gaobinlong) [reply src/plugins/data/public/query/query_string/dataset_service/dataset_service.ts]: Thanks, I've changed the code.

**Comment by ashwin-pc** [comment]
@gaobinlong we have merged ina few fixes since you raised this PR. IS this still needed?

**Comment by gaobinlong** [comment]
> @gaobinlong we have merged ina few fixes since you raised this PR. IS this still needed?

@ashwin-pc I've tested the newest main branch, the issue still exists, so this PR is still needed.

**Review by ruanyl — APPROVED** [review]

**Comment by ashwin-pc** [comment]
@gaobinlong Im not able to reproduce the issue originally linked in the PR. I do see the daatsource name missing, but the data itself is correct. Is that what you are seeing as well?

**Review by kavilla — APPROVED** [review]

**Comment by kavilla** [comment]
Great catch! Thank you @gaobinlong, going to merge in. Any feedback on the current dataset type config? Was it straight forward?  Or any issues you had while implementing.

**Comment by kavilla** [comment]
One thing to note: recent datasets could be stale so best way to recreate the original issue was to start in a fresh session set a default index to point to a remote cluster then navigate to Discover

**Pull request merged**